File: python/algos/is_happy.py
def digits_squared_sum(number: int) -> int:
    result = 0
    while number > 0:
        result += (number % 10) ** 2
        number //= 10
    return result


def is_happy(number: int) -> bool:
    seen = set()
    number = digits_squared_sum(number)
    while number != 1 and number not in seen:
        seen.add(number)
        number = digits_squared_sum(number)
    return number == 1


def test_is_happy():
    assert is_happy(19)
    assert not is_happy(2)

File: go/is_happy.go
package algos

func digitsSquaredSum(number int) int {
	result := 0
	for number > 0 {
		result += (number % 10) * (number % 10)
		number = number / 10
	}

	return result
}

func isHappy(number int) bool {
	seen := make(map[int]bool)
	for number != 1 {
		if _, ok := seen[number]; ok {
			break
		} else {
			seen[number] = true
			number = digitsSquaredSum(number)
		}
	}
	return number == 1
}
